project/app/src/test/java/org/owntracks/android/support/ParserTest.java

```diff
@@ -0,0 +1,187 @@
+package org.owntracks.android.support;
+
+import com.fasterxml.jackson.core.JsonParseException;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.owntracks.android.App;
+import org.owntracks.android.messages.MessageBase;
+import org.owntracks.android.messages.MessageLocation;
+import org.owntracks.android.messages.MessageUnknown;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.when;
+import static org.powermock.api.mockito.PowerMockito.mockStatic;
+
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({App.class})
+public class ParserTest {
+    private MessageLocation messageLocation;
+    @Mock
+    private
+    Preferences testPreferences;
+
+    @Before
+    public void setupMessageLocation() {
+        messageLocation = new MessageLocation();
+        messageLocation.setAcc(10);
+        messageLocation.setAlt(20);
+        messageLocation.setBatt(30);
+        messageLocation.setConn("TestConn");
+        messageLocation.setLat(50.1);
+        messageLocation.setLon(60.2);
+        messageLocation.setTst(123456789);
+    }
+
+    @Before
+    public void setupEncryptionProvider() {
+        mockStatic(App.class);
+        org.powermock.api.mockito.PowerMockito.when(App.getPreferences()).thenReturn(testPreferences);
+        when(testPreferences.getEncryptionKey()).thenReturn("testEncryptionKey");
+        when(encryptionProvider.isPayloadEncryptionEnabled()).thenReturn(true);
+    }
+
+    @Test
+    public void ParserCorrectlyConvertsLocationToPrettyJSON() throws Exception {
+        Parser parser = new Parser(null);
+        String expected = "{\n" +
```

Comment:
Thats a bit hard to read with all of the escaping, maybe better with triple quotes or stored on disk/loaded into a JSON parser?

Reply:
Fair point - I'll look into making it more readable.